src/CommandLine.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef COMMANDLINE_H
+#define COMMANDLINE_H
+
+#include "sbuf/SBuf.h"
+#include <vector>
+
+/// Manages arguments passed to the program, including the program name.
+/// The same info is passed to main() as argc and argv[] parameters.
+class CommandLine
+{
+public:
+    // codes for options without short option characters
+    enum LongCodes {
+        ForegroundCode = 1,
+        KidCode = 2
+    };
+
+    typedef std::pair<char, SBuf> OptionsPair;
+    typedef std::list<OptionsPair> Options;
+
+    CommandLine(int argc, char *argv[]);
+
+    /// \returns parsed kid option argument or an empty string
+    SBuf kidName() const;
+
+    /// apply all available command line options
+    void processOptions();
+
+    /// generate a new argument list from the parsed one,
+    /// supstituting argv[0] and adding/substituting kid option
+    const char **argv(const char *argv0, const char *kidName);
+
+    /// \returns Squid executable file name
+    SBuf execFile() const { return execFile_; }
+
+private:
+    void processOption(const char, const char *);
+    void parse(int argc, char *argv[]);
+
+    SBuf execFile_;
+    std::vector<const char *> argv_;
```

Comment:
It is not clear whether argv_ includes execFile_. It should be. If it does not include, we should document that unusual design because "normal" (i.e., main()'s) `argv[]` does . If it includes, we should probably remove the execFile_ member.

Reply:
The difficulty here is that we have to distinguish 'an executable file', i.e., the first argument of execvp() and  'argv[0]', which is what `ps` utility displays, but from within the main() we have an access only to argv[] array. Fortunately, we need only to start kids from master process, where argv[0] and this 'executable file' are the same (execFile_ is initialized from argv[0] as you noticed). For kids that would be wrong, but we do not need to spawn kids from kids.

Reply:
It sounds like you have (correctly) convinced yourself that there is no difficulty here :-). Everything you described is not CommandLine's problem. The fact that main.cc knows `argv[0]` meaning (in master context) is not a difficulty for the CommandLine class (or at all). The CommandLine class should not attempt to interpret `argv[0]` or assign some special meaning to its value. That's the class caller/user's job...